flow/diff_context.h

```diff
@@ -196,24 +196,29 @@ class DiffContext {
 
   Statistics& statistics() { return statistics_; }
 
+  SkRect MapRect(const SkRect& rect);
+
  private:
   struct State {
     State();
 
     bool dirty;
-    SkRect cull_rect;  // in screen coordinates
-
-    // In order to replicate paint process closely, we need both the original
-    // transform, and the overriden transform (set for layers that need to paint
-    // on integer coordinates). The reason for this is that during paint the
-    // transform matrix is overriden only after layer passes the cull check
-    // first (with original transform). So to cull layer we use transform, but
-    // to get paint coordinates we use transform_override. Child layers are
-    // painted after transform override, so if set we use transform_override as
-    // base when diffing child layers.
-    SkMatrix transform;
-    std::optional<SkMatrix> transform_override;
-    size_t rect_index_;
+
+    size_t rect_index;
+
+    // In order to replicate paint process closely, DiffContext needs to take
+    // into account that some layers are painted with transform translation
+    // snapped to integral coordinates.
+    //
+    // It's not possible to simply snap the transform itself, because culling
+    // needs to happen with original (unsnapped) transform, just like it does
+    // during paint. This means the integral coordinates must be applied after
```

Comment:
One complication is that a pixel snapped bounds may involve more than a single pixel of slop if it runs through an ImageFilter. That is potentially an issue for the accuracy of the bounds calculated in Preroll as well...

Reply:
Or an outset per time it happens, but it can happen on a number of layers leading to bounds bloat.

Reply:
I think the outset+1 in the quickReject is enough for Paint because it is only ever "1 pixel snap behind", but for Diff, it can have N nested pixel snaps...